File: src/engines/Counter.java
package engines;

import game.Board;
import game.Move;
import utils.MoveGenerator;

import java.util.LinkedHashMap;
import java.util.List;

public class Counter {
    private LinkedHashMap<String, Integer> results;
    private int depth;

    public Counter() {

    }

    public LinkedHashMap<String, Integer> goPerft(String fenString, int depth) {
        this.depth = depth;
        this.results = new LinkedHashMap<>();

        Board board = new Board(fenString);
        int numNodes = countPositions(board, depth);

        results.put("Nodes searched", numNodes);
        return results;
    }

    private int countPositions(Board board, int depth) {
        if (depth == 0) {
            return 1;
        }

        List<Move> moves = MoveGenerator.generateMoves(board);
        int numPositions = 0;

        for (Move move : moves) {
            board.makeMove(move);
            int movePositions = countPositions(board, depth - 1);
            board.undoMove(move);

            numPositions += movePositions;
            if (depth == this.depth) {
                System.out.println(move + ": " + movePositions);
                results.put(move.toString(), movePositions);
            }
        }

        return numPositions;
    }
}

File: src/counter.cpp
#include <iostream>

#include "counter.h"
#include "move.h"

std::unordered_map<std::string, int> Counter::goPerft(const std::string& fenString, int depth) {
    const auto board = new Board(fenString);

    const std::vector<Move> moves = MoveGenerator::generateMoves(*board);

    int numPositions = 0;
    for (const Move& move: moves) {
        if (depth == 1) {
            results[move.toString()] = 1;
            numPositions += 1;
            continue;
        }

        if (move.toString() == "f3f2") {
            std::cout << "here" << std::endl;
        }

        board->makeMove(move, true);
        const int movePositions = countPositions(*board, depth - 1);
        board->undoMove(move);

        results[move.toString()] = movePositions;
        numPositions += movePositions;
    }

    results["Nodes searched"] = numPositions;

    delete board;
    return results;
}

int Counter::countPositions(Board& board, const int depth) {
    const std::vector<Move> moves = MoveGenerator::generateMoves(board);

    //this is for bulk counting, no reason to actually make all these moves
    if (depth == 1) {
        return static_cast<int>(moves.size());
    }

    int numPositions = 0;
    for (const Move& move: moves) {
        board.makeMove(move, true);
        const int movePositions = countPositions(board, depth - 1);
        board.undoMove(move);

        numPositions += movePositions;
    }

    return numPositions;
}
